include/boost/spirit/home/x3/nonterminal/detail/rule.hpp

```diff
@@ -51,7 +51,7 @@ namespace boost { namespace spirit { namespace x3
     parse_rule(
         rule<ID, Attribute> rule_
       , Iterator& first, Iterator const& last
-      , Context const& context, ActualAttribute& attr);
+      , Context const& context, ActualAttribute& attr);    
```

Comment:
Will do.  My emacs editor, for some strange reason, puts those in and it's, obviously,
not immediately obvious  ;(

W.r.t. the rebase, I've been googling the question and read several web pages
and blogs but still don't have enough confidence to try it.  Currently I've followed
the `Merge Workflow` described here:

  https://randyfay.com/taxonomy/term/26

More specifically I've done:

> git pull upstream develop

Then manually resolved the CONFLICT's, then:

> run the revised test/x3/Jamfile

and make changes to get all tests to pass.  However, I still do not
understand, despite reading *many* webpages of `git rebase` docs,
 how *exactly* I should do a rebase to remove all these merge commits.
I'll continue to read such pages until I do understand or just
resort to trial and error trying, but it would speed things up
if you could tell me exactly how to do this rebase.

w.r.t. the strange files, they're not really needed, but they
were helpful to illustrate to the newsgroup readers solutions
to their problems.  I'll remove them (or maybe move them
to another repo).

Reply:
> More specifically I've done:
> > git pull upstream develop

Be aware that if you did not tweak git this command will produce a merge commit on your branch.

> how exactly I should do a rebase to remove all these merge commits.

Enter `git rebase upstream/develop` and solve merge conflicts. That's all. (I assume `upstream` is setup for boostorg repository). If you solve merge conflicts right they will disappear, otherwise they will turn into a regular commits (but you still can polish out them but squashing/removing/editing).

Reply:
Thanks Nikita.  I've replied to you via my email to:

  > notifications@github.com
  
In that email were details of how I tried your suggestion
without success. However, that reply is not turning up here;
so, this is to check that you're getting them.  If not, I'll
repeat those details here in another reply.

Reply:
I did not receive anything.

>  I tried your suggestion without success.

You failed to solve merge conflicts? (I don't think you've failed to type a single command). Well, since you've wrongly solved them currently too, do not bother then, just cherry-pick or re-do your patches on the latest develop (and then forcepush to this branch).

Reply:
Solving the merge conflicts was not the problem.  The problem was repeatedly solving
merge conflicts.  After the 1st `git rebase upstream/develop`, the git output listed the conflicts,
which were resolved.  Then, following the git output instructions, did `git rebase --continue`
which resulted in another set of merge conflicts.  After doing this 3-4 times, I googled and
found the -Xours option.  So, I used `git rebase --abort ; git rebase -Xours upstream/develop`.
This seemed to work, but then, when I tried `git push`, the error msg said:

> hint: Updates were rejected because the tip of your current branch is behind
> hint: its remote counterpart. Integrate the remote changes (e.g.
> hint: 'git pull ...') before pushing again.

Further googling lead to more suggestions for solving the problem but
they didn't work either.

So, I guess I'll do as you suggest with the latest develop.

Reply:
You'll have a heavy burden keeping up such a large PR with the moving code base. Please heed my request to keep the PRs minimal and incremental. Please focus on one small item at a time. Superfluous formatting also do not help as they tend to break the diffs.

That said, I am genuinely interested with your work.

Reply:
> when I tried `git push`, the error msg

This is a git protection of unintended rewriting of history. But you are rewriting it with a clear intention, that's why you need to force-push (`--force` flag).
If you need to save the previous branch history just make a copy of the branch.

Reply:
>  After doing this 3-4 times, I googled and found the -Xours option. So, I used `git rebase --abort ; git rebase -Xours upstream/develop` followed by only 1 set of merge conflict resolutions.

This is really the wrong way to do it. I swear you will solve merge conflicts in a bad way and will end with something you really do not want to.

Reply:
OK, Nikita.  I'll create 3 new branches to solve
3 separate problems, following Joel's suggestion, and cherry-pick,
as you suggest Nikita, from this branch to the 3 new branches.